core/tests/DesktopTests/AdalHeadlessAutomationTestApp/App/AuthenticationHelper.cs

```diff
@@ -0,0 +1,375 @@
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.Serialization.Json;
+using System.Threading.Tasks;
+using System.Web.Script.Serialization;
+using Microsoft.IdentityModel.Clients.ActiveDirectory.Internal;
+using Microsoft.IdentityModel.Clients.ActiveDirectory.Internal.Cache;
+
+// COPIED FROM ADAL.NET repo test automation app
+// https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/blob/dev/automation/WinFormsAutomationApp/AuthenticationHelper.cs
+
+namespace WinFormsAutomationApp
+{
+    internal static class AuthenticationHelper
+    {
+        #region Public Methods      
+        public static async Task<string> AcquireToken(Dictionary<string, string> input)
+        {
+            Dictionary<string, object> res = new Dictionary<string, object>();
+            AuthenticationContext ctx = new AuthenticationContext(input["authority"]);
+            try
+            {
+                AuthenticationResult result = null;
+
+                if (!input.ContainsKey("redirect_uri"))
+                {
+                    UserCredential userCred = new UserCredential();
+                    result = await ctx.AcquireTokenAsync(input["resource"], input["client_id"], userCred).ConfigureAwait(false);
+                }
+                else if (input.ContainsKey("user_identifier") && input.ContainsKey("password"))
+                {
+                    UserPasswordCredential user = new UserPasswordCredential(input["user_identifier"], input["password"]);
+                    result = await ctx.AcquireTokenAsync(input["resource"], input["client_id"], user).ConfigureAwait(false);
+                }
+                else if (input.ContainsKey("user_identifier") && input.ContainsKey("user_identifier_type"))
+                {
+                    UserIdentifierType userIdentifierType;
+                    UserIdentifierType.TryParse(input["user_identifier_type"], out userIdentifierType);
+                    string prompt = input.ContainsKey("prompt_behavior") ? input["prompt_behavior"] : null;
+                    result = await ctx.AcquireTokenAsync(input["resource"], input["client_id"], new Uri(input["redirect_uri"]),
+                        GetPlatformParametersInstance(prompt), 
+                        new UserIdentifier(input["user_identifier"], userIdentifierType))
+                        .ConfigureAwait(false);
+                }
+                else
+                {
+                    string prompt = input.ContainsKey("prompt_behavior") ? input["prompt_behavior"] : null;
+                    result = await ctx.AcquireTokenAsync(input["resource"], input["client_id"], new Uri(input["redirect_uri"]),
+                                GetPlatformParametersInstance(prompt)).ConfigureAwait(false);
+                }
+                res = ProcessResult(result, input);
+            }
+            catch (Exception exc)
+            {
+                res.Add("error", exc.Message);
```

Comment:
The same comment about using dictionaries applies here. Why can't we just propagate the errors?